File: C++/task#1.cpp
#include <iostream>
#include <vector>
#include <random>
#include <algorithm>
#include <map>

using namespace std;

////////////////////////////Task#1//////////////////////////////////////////////
void displayMatrix(vector<vector<int>> matrix) {
    if (matrix.empty() || matrix[0].empty()) {
        cout << "Матрица пуста!" << endl;
        return;
    }

    for (int i = 0; i < matrix.size(); i++) {
        for (int j = 0; j < matrix[i].size(); j++) {
            cout << matrix[i][j] << " ";
        }
        cout << endl;
    }
}

int inizialMatrix1(vector<vector<int>>& matrix, int por, mt19937& gen) {
    if (por % 2 == 1) {
        cout << "Матрица должна быть чётной, введите другое значение!" << endl;
        return 0;
    }

    matrix.clear();

    uniform_int_distribution<> dist(-50, 50);

    for (int row = 0; row < por; row++) {
        vector<int> str;
        for (int col = 0; col < por; col++) {
            int elem;
            if (col % 2 == 1) {
                do {
                    elem = dist(gen);
                } while (elem >= -10 && elem <= 10);
            } else {
                elem = dist(gen);
            }
            str.push_back(elem);
        }
        matrix.push_back(str);
    }

    return 1;
}
////////////////////////////////////////////////////////////////////////////////

////////////////////////////Task#2//////////////////////////////////////////////
int inizialMatrix2(vector<vector<int>>& matrix, int str, int col, mt19937& gen) {
    uniform_int_distribution<> dist(-100, 100);

    matrix.clear(); // на случай, если матрица уже была заполнена

    for (int i = 0; i < str; i++) {
        vector<int> row;
        for (int j = 0; j < col; j++) {
            int elem = dist(gen);
            row.push_back(elem);
        }
        matrix.push_back(row);
    }

    return 1;
}

int modifiedMatrix(vector<vector<int>>& matrix) {
    vector<pair<int, int>> pos_neg;
    for (const auto& row : matrix) {
        int pos = 0, neg = 0;
        for (int val : row) {
            if (val > 0) pos++;
            else if (val < 0) neg++;
        }
        pos_neg.push_back({pos, neg});
    }

    int k = 0;
    for (const auto& pair : pos_neg) {
        if (pair.first > pair.second) {
            auto middle = matrix[k].begin() + matrix[k].size()/2;
            sort(middle, matrix[k].end());
            reverse(middle, matrix[k].end());
            k += 1;
        }
        if (pair.first < pair.second) {
            for (int i = 0; i < matrix[k].size(); i++) {
                matrix[k][i] = matrix[k][i] * matrix[k][i];
            }
            k += 1;
        }
        if (pair.first == pair.second) {
            for (int i = 0; i < matrix[k].size(); i++) {
                matrix[k][i] = 0;
            }
            k += 1;
        }
    }

    return 1;
}

void displayMatrix2(vector<vector<int>> matrix, int col, int row) {
    if (matrix.empty() || matrix[0].empty()) {
        cout << "Матрица пуста!" << endl;
        return;
    }

    string dlina = "------";
    string resstr = "";
    for (int i = 0; i < row; i++) {
        resstr += dlina;
    }

    for (int j = col - 1; j > -1; j--) {
        for (int i = 0; i < row; i++) {
            cout << matrix[i][j];
            if (i != row - 1) {
                cout << " | ";
            }
        }
        if (j != 0) {
            cout << endl << resstr << endl;
        } 
    }
    cout << endl;
}
////////////////////////////////////////////////////////////////////////////////

////////////////////////////Task#3//////////////////////////////////////////////
map<int, int> subMatrix(vector<vector<int>>& matrix, int str, int col, mt19937& gen) {
    map<int, int> result;

    int maxSize = min(str, col);
    for (int k = 1; k <= maxSize; ++k) {
        int count = 0;
        // Перебираем все возможные позиции верхнего левого угла подматрицы k×k
        for (int i = 0; i <= str - k; ++i) {
            for (int j = 0; j <= col - k; ++j) {
                bool allOne = true;
                for (int x = 0; x < k && allOne; ++x) {
                    for (int y = 0; y < k; ++y) {
                        if (matrix[i + x][j + y] != 1) {
                            allOne = false;
                            break;
                        }
                    }
                }
                if (allOne) {
                    ++count;
                }
            }
        }
        result[k] = count;
    }

    return result;
}
////////////////////////////////////////////////////////////////////////////////

int main() {
    random_device rd;
    mt19937 gen(rd());

    while (true) {
        cout << "Введите номер задания: ";
        string task;
        cin >> task;
        if (task == "1") {
            vector<vector<int>> matrix;
            while(true) {
                cout << "Напишите команду: ";
                string str;
                cin >> str;
                if (str == "create") {
                    int n;
                    while (true) {
                        cout << "Введите порядок матрицы: ";
                        cin >> n;
                        int res = inizialMatrix1(matrix, n, gen);
                        if (res == 1) {
                            break;
                        }
                    }
                    cout << "Матрица заполнена случайными числами от -50 до 50(за исключением чётных столбцов(в них нет чисел из диапазона [-10, 10]))\n";
                } 
                if (str == "print") {
                    displayMatrix(matrix);
                }
                if (str == "exit") {
                    break;
                }
            }
        }
        if (task == "2") {
            vector<vector<int>> matrix;
            int row, col;
            cout << "Введите длину матрицы(M): ";
            cin >> row;
            cout << "Введите ширину матрицы(N): ";
            cin >> col;
            int res = inizialMatrix2(matrix, row, col, gen);
            cout << "Сгенерированная матрица(MxN):\n";
            displayMatrix2(matrix, col, row);
            cout << endl;

            cout << "Модифицированная матрица:\n";
            res = modifiedMatrix(matrix);
            displayMatrix2(matrix, col, row);
            cout << endl;
        }
        if (task == "3") {
            vector<vector<int>> matrix = {
                {1, 1, 1, 0},
                {1, 1, 1, 1},
                {1, 1, 1, 1},
                {0, 1, 1, 0}
            };
            
            map<int, int> res = subMatrix(matrix, 4, 4, gen);
            cout << "Матрица:" << endl;
            displayMatrix(matrix);
            cout << endl;

            int k = 0;
            for (const auto& i : res) {
                k += i.second;
            }

            cout << "Всего найдено " << k << " подматриц:" << endl;

            int i = 0;
            for (const auto& i : res) {
                if (i.second != 0) {
                    cout << i.second << " подматри(ц/цы/а) со стороной " << i.first << endl;
                }
            }
            cout << endl;
        }
        if (task == "exit") {
            break;
        }
    }
}

File: Python/task#1.py
import random

############################ Task#1 ##############################
def display_matrix(matrix):
    if not matrix or not matrix[0]:
        print("Матрица пуста!")
        return

    for row in matrix:
        print(" ".join(map(str, row)))

def inizial_matrix_1(por):
    if por % 2 == 1:
        print("Матрица должна быть чётной, введите другое значение!")
        return None

    matrix = []
    for row in range(por):
        current_row = []
        for col in range(por):
            if col % 2 == 1:
                # выбираем случайное число вне диапазона [-10,10]
                while True:
                    elem = random.randint(-50, 50)
                    if elem < -10 or elem > 10:
                        break
            else:
                elem = random.randint(-50, 50)
            current_row.append(elem)
        matrix.append(current_row)

    return matrix
##################################################################

############################ Task#2 ##############################
def inizial_matrix_2(rows, cols):
    matrix = []
    for _ in range(rows):
        row = [random.randint(-100, 100) for _ in range(cols)]
        matrix.append(row)
    return matrix

def modified_matrix(matrix):
    for i, row in enumerate(matrix):
        pos = sum(1 for val in row if val > 0)
        neg = sum(1 for val in row if val < 0)

        if pos > neg:
            middle = len(row) // 2
            # сортируем правую половину по убыванию
            right_part = sorted(row[middle:], reverse=True)
            matrix[i] = row[:middle] + right_part

        elif pos < neg:
            # возводим элементы строки в квадрат
            matrix[i] = [val * val for val in row]

        else:  # pos == neg
            matrix[i] = [0] * len(row)

def display_matrix_2(matrix, cols, rows):
    if not matrix or not matrix[0]:
        print("Матрица пуста!")
        return

    dlina = "------"
    resstr = dlina * rows

    for j in reversed(range(cols)):
        for i in range(rows):
            print(matrix[i][j], end="")
            if i != rows - 1:
                print(" | ", end="")
        if j != 0:
            print("\n" + resstr)
    print()
##################################################################

############################ Task#3 ##############################
def sub_matrix(matrix, rows, cols):
    result = {}
    max_size = min(rows, cols)

    for k in range(1, max_size + 1):
        count = 0
        for i in range(rows - k + 1):
            for j in range(cols - k + 1):
                all_one = True
                for x in range(k):
                    for y in range(k):
                        if matrix[i + x][j + y] != 1:
                            all_one = False
                            break
                    if not all_one:
                        break
                if all_one:
                    count += 1
        result[k] = count
    return result
##################################################################

def main():
    while True:
        task = input("Введите номер задания: ")
        if task == "1":
            matrix = None
            while True:
                command = input("Напишите команду: ")
                if command == "create":
                    while True:
                        n = int(input("Введите порядок матрицы: "))
                        matrix = inizial_matrix_1(n)
                        if matrix is not None:
                            break
                    print("Матрица заполнена случайными числами от -50 до 50(за исключением чётных столбцов(в них нет чисел из диапазона [-10, 10]))")
                elif command == "print":
                    if matrix:
                        display_matrix(matrix)
                    else:
                        print("Матрица не создана.")
                elif command == "exit":
                    break

        elif task == "2":
            row = int(input("Введите длину матрицы(M): "))
            col = int(input("Введите ширину матрицы(N): "))
            matrix = inizial_matrix_2(row, col)
            print("Сгенерированная матрица(MxN):")
            display_matrix_2(matrix, col, row)
            print()

            modified_matrix(matrix)
            print("Модифицированная матрица:")
            display_matrix_2(matrix, col, row)
            print()

        elif task == "3":
            matrix = [
                [1, 1, 1, 0],
                [1, 1, 1, 1],
                [1, 1, 1, 1],
                [0, 1, 1, 0]
            ]
            res = sub_matrix(matrix, 4, 4)
            print("Матрица:")
            display_matrix(matrix)
            print()

            total = sum(res.values())
            print(f"Всего найдено {total} подматриц:")

            for k, count in res.items():
                if count != 0:
                    print(f"{count} подматри(ц/цы/а) со стороной {k}")
            print()

        elif task == "exit":
            break

if __name__ == "__main__":
    main()
